### PR TITLE
Support functions for message in driver.wait()

### PR DESCRIPTION
**Pull request by jackall3n**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Description
This change allows users to pass a function that resolves an error message when the `driver.wait()` condition fails. Currently a `string` can be passed to the `wait()` method, now `() => string` can also be passed.

### Motivation and Context
Currently a user only has the ability to pass a message containing information about the step at the beginning of the test. Giving a user the ability to pass a function that resolves a message allows the message to contain information that is made available at the end of the test. The motivation for this PR is to be able to write out information about the current state of the test results after a timeout period has expired.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by jackall3n** [comment]
Hey @shs96c, I've signed the CLA provided above.

File: javascript/node/selenium-webdriver/lib/webdriver.js [hunk 1]
```diff
@@ -786,11 +786,8 @@ class WebDriver {
         let start = Date.now();
         let timer = setTimeout(function() {
           timer = null;
-          reject(
-              new error.TimeoutError(
-                  (message ? `${message}\n` : '')
-                      + 'Timed out waiting for promise to resolve after '
-                      + (Date.now() - start) + 'ms'));
+          let timeoutMessage = message ? `${typeof message === 'function' ? message() : message}\n` : '';
```
Comment (jleyba) [inline javascript/node/selenium-webdriver/lib/webdriver.js]: If the message function throws, the promise will never be resolved. So you probably need something like:

```
try {
  let msg = message();  // Shortened for brevity
  reject(new error.TimeoutError(msg));
} catch (ex) {
  reject(new error.TimeoutError(`${ex}\nTimed out waiting...`));
}
```

Comment (jackall3n) [reply javascript/node/selenium-webdriver/lib/webdriver.js]: @jleyba good idea. updated the PR with additional tests for these scenarios.

File: javascript/node/selenium-webdriver/lib/webdriver.js [hunk 2]
```diff
@@ -786,11 +786,13 @@ class WebDriver {
...
-                  (message ? `${message}\n` : '')
-                      + 'Timed out waiting for promise to resolve after '
-                      + (Date.now() - start) + 'ms'));
+          try {
+            let timeoutMessage = message ? `${typeof message === 'function' ? message() : message}\n` : '';
+            reject(new error.TimeoutError(`${timeoutMessage}Timed out waiting for promise to resolve after ${Date.now() - start}ms`));
+          }
+          catch (ex) {
+            reject(new error.TimeoutError(`${ex.message}\nTimed out waiting for promise to resolve after ${Date.now() - start}ms`));
+          }
```
Comment (jleyba) [inline javascript/node/selenium-webdriver/lib/webdriver.js]: This could be extracted to a function (since it's basically repeated below)

Comment (jackall3n) [reply javascript/node/selenium-webdriver/lib/webdriver.js]: Agreed, have you got a preferred directory for this function to live? Or within the same file?

Comment (jackall3n) [reply javascript/node/selenium-webdriver/lib/webdriver.js]: @jleyba I've extracted the function out

Comment (jackall3n) [reply javascript/node/selenium-webdriver/lib/webdriver.js]: @jleyba anything more you can think of adding?

**Comment by jackall3n** [comment]
Any thoughts on this one?

**Pull request closed without merging**